sql/analyzer/resolve_columns.go

```diff
@@ -183,34 +236,524 @@ func dedupStrings(in []string) []string {
+			// a table lower in the tree must provide the source
+			s.addCols(baseCol)
+		}
+	}
+	for t := range outer.stars {
+		if t == alias {
+			s.stars[table] = struct{}{}
+		}
+	}
+	s.unqualifiedStar = s.unqualifiedStar || outer.unqualifiedStar
+}
+
+// addColsFromNode adds all of a node's column dependencies to this scope
+func (s *scope) addColsFromNode(n sql.Node) {
+	ne, ok := n.(sql.Expressioner)
+	if !ok {
+		return
+	}
+	for _, e := range ne.Expressions() {
+		transform.InspectExpr(e, func(e sql.Expression) bool {
+			switch e := e.(type) {
+			case *expression.Alias:
+				switch e := e.Child.(type) {
+				case *expression.GetField:
+					s.addCols(tableCol{table: strings.ToLower(e.Table()), col: strings.ToLower(e.Name())})
+				case *expression.UnresolvedColumn:
+					s.addCols(tableCol{table: strings.ToLower(e.Table()), col: strings.ToLower(e.Name())})
+				default:
+				}
+			case *expression.GetField:
+				s.addCols(tableCol{table: strings.ToLower(e.Table()), col: strings.ToLower(e.Name())})
+			case *expression.UnresolvedColumn:
+				s.addCols(tableCol{table: strings.ToLower(e.Table()), col: strings.ToLower(e.Name())})
+			case *expression.Star:
+				if len(e.Table) > 0 {
+					s.stars[strings.ToLower(e.Table)] = struct{}{}
+				} else {
+					s.unqualifiedStar = true
+				}
+			default:
+			}
+			return false
+		})
+	}
+}
+
+type builder struct {
+	ctx        *sql.Context
+	root       sql.Node
+	tableCols  map[sql.RelId][]tableCol
+	id         sql.RelId
+	inSubquery string
+}
+
+func newBuilder(ctx *sql.Context, root sql.Node) *builder {
+	return &builder{
+		ctx:       ctx,
+		root:      root,
+		tableCols: make(map[sql.RelId][]tableCol),
+	}
+}
+
+func compareTableCol(i, j tableCol) int {
+	if i.table < j.table {
+		return -1
+	} else if i.table > j.table {
+		return 1
+	} else if i.col < j.col {
+		return -1
+	} else if i.col > j.col {
+		return 1
+	}
+	return 0
+}
+
+// mergeTableCols combines two sets of []tableCol, carefully maintaining
+// ordering to keep the projection outputs deterministic.
+func mergeTableCols(x, y []tableCol) []tableCol {
+	if len(x) == 0 {
+		return y
+	} else if len(y) == 0 {
+		return x
+	}
+	newCols := make([]tableCol, len(x)+len(y))
+	var i, j, k int
+	for i < len(x) && j < len(y) {
+		cmp := compareTableCol(x[i], y[j])
+		if cmp < 0 {
+			newCols[k] = x[i]
+			i++
+		} else if cmp > 0 {
+			newCols[k] = y[j]
+			j++
+		} else {
+			newCols[k] = x[i]
+			i++
+			j++
+		}
+		k++
+	}
+	for i < len(x) {
+		newCols[k] = x[i]
+		i++
+		k++
+	}
+	for j < len(y) {
+		newCols[k] = y[j]
+		j++
+		k++
+	}
+	return newCols[:k]
+}
+
+// addTableCols combines the scope dependencies, maintaining schema sort ordering
+func (b *builder) addTableCols(id sql.RelId, other []tableCol) {
+	curr, _ := b.tableCols[id]
+	b.tableCols[id] = mergeTableCols(curr, other)
+}
+
+func (b *builder) delTableCols(id sql.RelId) {
+	delete(b.tableCols, id)
+}
+
+func (b *builder) buildTable(n sql.Node, inScope *scope) *scope {
```

Comment:
Reconsider this name, not obvious what this method should do from name

Reply:
well, in theory this should collect applicable filters and projections, build a `*plan.ResolvedTable`, lift sourced columns recursively upwards to qualify parent dependencies, and then stick this relation in a memo. I would like it to do that at some point.